### ISSUE TITLE
Question about `safe_set_local_grad` and `safe_get_local_grad`

### ISSUE THREAD

**Issue by breaddaerb** [deepspeedai/DeepSpeed#7292]
deepspeed==0.16.7
I'm using zero stage-3

I write a custom Trainer to do some custom grad clipping after `backward()` is called and before `step()`. I use `safe_get_local_grad` and `safe_set_local_grad` to do the custom grad clipping. After clipping, I call `safe_get_local_grad` again, and the result is as expected. But then i add a callback on the end of `step()` and print the `ds_engine.optimizer._global_grad_norm`, it equals to the original grad(before the clipping). Does it mean I fail to set new value to grad?

**Comment by breaddaerb** [comment]
I take a closer look at `DeepSpeedZeroOptimizer_Stage3`. The `_global_grad_norm` is calculated in the `step()` by calling ` _get_norm_groups()` first, and it takes `averaged_gradients` and `fp16_groups` to get norm groups.
https://github.com/deepspeedai/DeepSpeed/blob/b048cc2b46a559b8f81c4315008cec34e650bbe0/deepspeed/runtime/zero/stage3.py#L2095
https://github.com/deepspeedai/DeepSpeed/blob/b048cc2b46a559b8f81c4315008cec34e650bbe0/deepspeed/runtime/zero/stage3.py#L1887
Does it mean the gradient is not changed at all or just the `_global_grad_norm` is not updated with the new grad?

**Comment by sfc-gh-truwase** [comment]
@breaddaerb, since multiple data structures are involved it is hard to say without repro code. Are you able to provide some test code?

**Comment by therealnaveenkamal** [comment]
Hi @breaddaerb / @sfc-gh-truwase,

I've written a piece of code to repro this.

```
import torch
import torch.nn as nn
import deepspeed
from deepspeed.utils import safe_get_local_grad, safe_set_local_grad
import argparse

class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = nn.Linear(4, 8)
        self.lin2 = nn.Linear(8, 2)

    def forward(self, x):
        x = self.lin1(x)
        x = self.lin2(x)
        return x

def get_deepspeed_config():
    return {
        "train_batch_size": 8,
        "gradient_clipping": 1.0,
        "zero_optimization": {
            "stage": 3,
            "offload_optimizer": {"device": "cpu"},
            "overlap_comm": False,
            "contiguous_gradients": True,
            "reduce_bucket_size": 5e7,
            "stage3_prefetch_bucket_size": 5e7,
            "stage3_param_persistence_threshold": 1e4
        },
        "fp16": {"enabled": True, "initial_scale_power": 12},
        "gradient_accumulation_steps": 1,
        "steps_per_print": 1,
        "optimizer": {
            "type": "Adam",
            "params": {
                "lr": 1e-3,
                "betas": [0.9, 0.999],
                "eps": 1e-8,
                "weight_decay": 0
            }
        }
    }

def custom_grad_clip(model_engine, clip_value=0.1):
    modified_count = 0
    
    for param in model_engine.parameters():
        if not hasattr(param, 'ds_id'):
            continue
            
        grad = safe_get_local_grad(param)
        if grad is not None:
            # Apply clipping
            clamped_grad = torch.clamp(grad, -clip_value, clip_value)
            safe_set_local_grad(param, clamped_grad)
            modified_count += 1
    
    return modified_count

def get_grad_norms(model_engine):
    try:
        norm_groups = model_engine.optimizer._get_norm_groups()
        global_norm = torch.linalg.vector_norm(torch.stack(norm_groups))

        return global_norm, norm_groups, model_engine.optimizer._global_grad_norm
    except Exception as e:
        print(f"Warning: Could not get gradient norms: {e}")
        return None, None

def main():
    # Initialize model and DeepSpeed
    model = SimpleModel()
    parser = argparse.ArgumentParser()
    parser.add_argument('--local_rank', type=int, default=-1, help="Local rank for distributed training")
    args, _ = parser.parse_known_args()

    # Initialize DeepSpeed
    model_engine, optimizer, _, _ = deepspeed.initialize(
        args=args,
        model=model,
        config=get_deepspeed_config()
    )

    model_engine.train()

    # Create dummy data
    torch.manual_seed(10)
    inputs = torch.randn(8, 4, device=model_engine.device)
    targets = torch.randn(8, 2, device=model_engine.device)

    if model_engine.fp16_enabled():
        inputs = inputs.half()
        targets = targets.half()

    # Forward pass
    outputs = model_engine(inputs)
    loss = nn.MSELoss()(outputs, targets)
    print(f"Initial loss: {loss.item():.6f}")

    # Backward pass
    model_engine.backward(loss)

    # Get initial global norm
    initial_global_norm = model_engine.optimizer._global_grad_norm
    print(f"Initial global grad norm: {initial_global_norm:.6f}")

    # Get norm groups before clipping
    pre_clip_global_norm, pre_clip_norm_groups, pre_clip_global_grad_norm = get_grad_norms(model_engine)
    if pre_clip_global_norm is not None:
        print(f"Pre-clipping global norm: {pre_clip_global_norm:.6f}")

    # Apply custom gradient clipping
    modified_count = custom_grad_clip(model_engine, clip_value=0.1)
    print(f"Modified {modified_count} parameter gradients")

    # Get norm groups after clipping
    post_clip_global_norm, post_clip_norm_groups, post_clip_global_grad_norm = get_grad_norms(model_engine)
    if post_clip_global_norm is not None:
        print(f"Post-clipping global norm: {post_clip_global_norm:.6f}")

    # Run optimizer step
    model_engine.step()
    print("Optimizer step completed successfully")

    final_global_norm = model_engine.optimizer._global_grad_norm


    # Print summary
    print("\nSummary:")
    if pre_clip_norm_groups is not None:
        print(f"Pre-clipping norms: {[norm.item() for norm in pre_clip_norm_groups]}")
    if post_clip_norm_groups is not None:
        print(f"Post-clipping norms: {[norm.item() for norm in post_clip_norm_groups]}")


    print(f"Pre-clipping global grad norm: {pre_clip_global_grad_norm:.6f}")
    print(f"Post-clipping global grad norm: {post_clip_global_grad_norm:.6f}")
    print(f"Final global grad norm: {final_global_norm:.6f}")


if __name__ == "__main__":
    main()
```


**When CPU Offloading is True**
```
Initial loss: 0.859863
Initial global grad norm: 0.000000
Pre-clipping global norm: 5026.077637
Modified 4 parameter gradients
Post-clipping global norm: 5026.077637
[2025-05-21 20:30:27,023] [INFO] [logging.py:107:log_dist] [Rank 0] step=1, skipped=0, lr=[0.001], mom=[[0.9, 0.999]]
Optimizer step completed successfully

Summary:
Pre-clipping norms: [5026.07763671875]
Post-clipping norms: [5026.07763671875]
Pre-clipping global grad norm: 0.000000
Post-clipping global grad norm: 0.000000
**Final global grad norm: 1.227070**
```

**No CPU Offloading**
```
Initial loss: 0.832031
Initial global grad norm: 0.000000
Pre-clipping global norm: 4935.634191
Modified 4 parameter gradients
Post-clipping global norm: 0.761391
[2025-05-21 20:29:35,934] [INFO] [logging.py:107:log_dist] [Rank 0] step=1, skipped=0, lr=[0.001], mom=[[0.9, 0.999]]
Optimizer step completed successfully

Summary:
Pre-clipping norms: [4935.634190913677]
Post-clipping norms: [0.7613913786257984]
Pre-clipping global grad norm: 0.000000
Post-clipping global grad norm: 0.000000
**Final global grad norm: 0.000186**
```

**Problem**
Have a look at `Final global grad norm` in the outputs.
The `averaged_gradients` is not updated when CPU offloading is enabled
https://github.com/deepspeedai/DeepSpeed/blob/b048cc2b46a559b8f81c4315008cec34e650bbe0/deepspeed/runtime/zero/stage3.py#L1147

Thats why even though gradients are clipped, `_global_grad_norm` still uses the old averaged_gradients. If we change ds_config to `"offload_optimizer": {"device": "none"}`, `_global_grad_norm` will be updated.


For CPU offloading, gradients are already in `fp32_partitioned_groups_flat.grad`. Can raise a PR for this. Please let me know your thoughts @sfc-gh-truwase @tjruwase

**Comment by sfc-gh-truwase** [comment]
@therealnaveenkamal, awesome work. It looks like you really nailed the issue. A PR will be greatly appreciated. Thanks so much.

**Comment by breaddaerb** [comment]
@therealnaveenkamal Thanks for your analysis! However, when I use `offload_optimizer": {"device": "none"}`, the `_global_grad_norm` is still not updated. Is this behavior potentially related to i use bf16 mixed precision training?

**Comment by therealnaveenkamal** [comment]
@sfc-gh-truwase Thanks! I've raised a PR for the same.

@breaddaerb I tried with BF16 enabled and it works fine for me. Would you be able to provide your outputs for the above code? What is the value of your gradient_norms and the clip_value?

@

**Comment by breaddaerb** [comment]
Thank you so much for your help! @therealnaveenkamal  It turns out I made a mistake when implementing your solution earlier. After double-checking, your approach works perfectly! 

I really appreciate you taking the time to assist me with this issue. Closing this now as resolved.
